File: step_08_bit_manipulation/8.1_learn_bits/P008_swap_two_numbers/Solution.java
/**
 * Problem: Swap Two Numbers (Without Temp Variable)
 * Difficulty: EASY | XP: 10
 *
 * Swap two numbers without using a temporary variable.
 *
 * @author DSA_Nova
 */

// ============================================================
// Approach 1: Brute Force (Using Temp Variable)
// Time: O(1) | Space: O(1)
// ============================================================
class BruteForce {
    public static int[] swap(int a, int b) {
        int temp = a;
        a = b;
        b = temp;
        return new int[]{a, b};
    }
}

// ============================================================
// Approach 2: Optimal (XOR Swap)
// Time: O(1) | Space: O(1) -- no extra variable
// ============================================================
class Optimal {
    public static int[] swap(int a, int b) {
        a = a ^ b;  // a now holds a XOR b
        b = a ^ b;  // b = (a XOR b) XOR b = original a
        a = a ^ b;  // a = (a XOR b) XOR a = original b
        return new int[]{a, b};
    }
}

// ============================================================
// Approach 3: Best (Arithmetic Swap -- No XOR)
// Time: O(1) | Space: O(1)
// NOTE: Works for ints; beware of overflow with very large values.
// ============================================================
class Best {
    public static int[] swap(int a, int b) {
        a = a + b;
        b = a - b;  // b = (a + b) - b = original a
        a = a - b;  // a = (a + b) - a = original b
        return new int[]{a, b};
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Swap Two Numbers ===\n");

        int[][] tests = {
            {5, 10}, {0, 7}, {-3, 4}, {100, 100}, {0, 0}, {-5, -8}
        };

        for (int[] t : tests) {
            int a = t[0], b = t[1];
            int[] bf = BruteForce.swap(a, b);
            int[] op = Optimal.swap(a, b);
            int[] bs = Best.swap(a, b);

            boolean pass = bf[0] == b && bf[1] == a
                        && op[0] == b && op[1] == a
                        && bs[0] == b && bs[1] == a;

            System.out.println("Input: a=" + a + ", b=" + b);
            System.out.println("  Brute:   [" + bf[0] + ", " + bf[1] + "]");
            System.out.println("  Optimal: [" + op[0] + ", " + op[1] + "]");
            System.out.println("  Best:    [" + bs[0] + ", " + bs[1] + "]");
            System.out.println("  " + (pass ? "PASS" : "FAIL"));
            System.out.println();
        }
    }
}

File: step_08_bit_manipulation/8.1_learn_bits/P008_swap_two_numbers/solution.py
"""
Problem: Swap Two Numbers (Without Temp Variable)
Difficulty: EASY | XP: 10

Swap two numbers without using a temporary variable.
"""
from typing import Tuple


# ============================================================
# APPROACH 1: BRUTE FORCE (Using Temp Variable)
# Time: O(1)  |  Space: O(1)
# ============================================================
def brute_force(a: int, b: int) -> Tuple[int, int]:
    """Classic swap using a temporary variable."""
    temp = a
    a = b
    b = temp
    return a, b


# ============================================================
# APPROACH 2: OPTIMAL (XOR Swap)
# Time: O(1)  |  Space: O(1) -- no extra variable
# ============================================================
def optimal(a: int, b: int) -> Tuple[int, int]:
    """XOR swap trick: a^=b; b^=a; a^=b."""
    a = a ^ b  # a now holds a XOR b
    b = a ^ b  # b = (a XOR b) XOR b = a (original)
    a = a ^ b  # a = (a XOR b) XOR a = b (original)
    return a, b


# ============================================================
# APPROACH 3: BEST (Pythonic Tuple Swap)
# Time: O(1)  |  Space: O(1)
# ============================================================
def best(a: int, b: int) -> Tuple[int, int]:
    """Python one-liner using tuple unpacking."""
    a, b = b, a
    return a, b


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Swap Two Numbers ===\n")

    test_cases = [
        (5, 10),
        (0, 7),
        (-3, 4),
        (100, 100),
        (0, 0),
        (-5, -8),
        (2147483647, -2147483648),
    ]

    for a, b in test_cases:
        b1 = brute_force(a, b)
        o1 = optimal(a, b)
        r1 = best(a, b)
        expected = (b, a)
        status = "PASS" if b1 == expected and o1 == expected and r1 == expected else "FAIL"
        print(f"Input: a={a}, b={b}")
        print(f"  Brute:    {b1}")
        print(f"  Optimal:  {o1}")
        print(f"  Best:     {r1}")
        print(f"  Expected: {expected}  [{status}]\n")
